File: backend/src/modules/risk/domain/risk.ts
import { v4 as uuidv4 } from 'uuid';

export type RiskInput = {
  challengeId: string;
  startingBalance: number;
  equity: number;
  highWaterMark: number;
  dayPnl: number;
  tradingDays: number;
  profitTargetPct: number;
  dailyLossPct: number;
  maxLossPct: number;
  minTradingDays: number;
};

export type RiskResult =
  | { kind: 'ok' }
  | { kind: 'breach'; rule: string; reason: string }
  | { kind: 'target'; equity: number; tradingDays: number };

export class BreachRecord {
  constructor(
    public id: string,
    public challengeId: string,
    public rule: string,
    public reason: string,
    public createdAt: Date,
  ) {}

  static create(challengeId: string, rule: string, reason: string) {
    return new BreachRecord(uuidv4(), challengeId, rule, reason, new Date());
  }
}

/** Pure risk rules — no I/O (US-F1..F3). */
export function evaluateRisk(input: RiskInput): RiskResult {
  const dailyLimit = input.startingBalance * (input.dailyLossPct / 100);
  if (input.dayPnl <= -dailyLimit) {
    return {
      kind: 'breach',
      rule: 'DailyLoss',
      reason: `Daily loss ${input.dayPnl.toFixed(2)} exceeded -${dailyLimit.toFixed(2)}`,
    };
  }
  const maxLossFloor = input.startingBalance * (1 - input.maxLossPct / 100);
  if (input.equity < maxLossFloor) {
    return {
      kind: 'breach',
      rule: 'MaxDrawdown',
      reason: `Equity ${input.equity.toFixed(2)} below floor ${maxLossFloor.toFixed(2)}`,
    };
  }
  const target = input.startingBalance * (1 + input.profitTargetPct / 100);
  if (input.equity >= target && input.tradingDays >= input.minTradingDays) {
    return { kind: 'target', equity: input.equity, tradingDays: input.tradingDays };
  }
  if (input.equity >= target && input.tradingDays < input.minTradingDays) {
    return { kind: 'ok' };
  }
  return { kind: 'ok' };
}

File: backend-python/app/modules/risk/domain/risk.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class RiskInput:
    starting_balance: float
    equity: float
    day_pnl: float
    trading_days: int
    profit_target_pct: float
    daily_loss_pct: float
    max_loss_pct: float
    min_trading_days: int


@dataclass(frozen=True)
class RiskResult:
    kind: str  # ok | breach | target
    rule: str = ""
    reason: str = ""


def evaluate_risk(inp: RiskInput) -> RiskResult:
    daily_limit = inp.starting_balance * (inp.daily_loss_pct / 100)
    if inp.day_pnl <= -daily_limit:
        return RiskResult(
            kind="breach",
            rule="DailyLoss",
            reason=f"Daily loss {inp.day_pnl:.2f} exceeded -{daily_limit:.2f}",
        )
    max_loss_floor = inp.starting_balance * (1 - inp.max_loss_pct / 100)
    if inp.equity < max_loss_floor:
        return RiskResult(
            kind="breach",
            rule="MaxDrawdown",
            reason=f"Equity {inp.equity:.2f} below floor {max_loss_floor:.2f}",
        )
    target = inp.starting_balance * (1 + inp.profit_target_pct / 100)
    if inp.equity >= target and inp.trading_days >= inp.min_trading_days:
        return RiskResult(kind="target")
    return RiskResult(kind="ok")
